study1/week3/reduce-array-size-to-the-half.java

```diff
@@ -0,0 +1,30 @@
+// https://leetcode.com/problems/reduce-array-size-to-the-half/
+// Time Complexity: O(n), n: arr.length
+// Space Complexity: O(n), n: arr.length
+// In order that at least half of the integers of the array are removed, the most numerous integer has to be removed first (The more integer are removed, the faster can the array be smaller)
+class Solution {
+    public int minSetSize(int[] arr) {
+        int setSize = 0;
+        HashMap<Integer, Integer> integerMap = new HashMap<>();
+
+        // put value into map
+        for (int a : arr) {
+            integerMap.put(a, integerMap.getOrDefault(a, 0) + 1);
+        }
+
+        // in a descending order
+        List<Integer> keyList = new ArrayList<>(integerMap.keySet());
+        Collections.sort(keyList, (value1, value2) -> (integerMap.get(value2).compareTo(integerMap.get(value1))));
+
+        int cntInteger = 0; // for counting the number of integers which are removed
+        for (int key : keyList) {
+            cntInteger += integerMap.get(key);
+            setSize++;
+            if (cntInteger >= arr.length/2) {
```

Comment:
So there is a separate termination condition of this loop, meaning that it can be best represented as a while loop.
```java
while (cntInteger < arr.length/2) {
   int key = keyList.get(setSize);
   cntInteger += integerMap.get(key);
   setSize++;
}
```
You don't have to worry that this loop will run forever and be an infinite loop, because we will at one point reach the half of the array(in the best after one index, in the worst really past the half of the indicies), so this loop will always terminate.